presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -561,6 +549,86 @@ private void invokeNoMoreSplitsIfNecessary()
         return COMPLETED_FUTURE;
     }
 
+    private Iterator<InternalHiveSplit> buildManifestFileIterator(Table table,
+            FileStatus symlinkFile,
+            FileSystem fileSystem,
+            DirectoryLister directoryLister,
+            NamenodeStats namenodeStats,
+            InternalHiveSplitFactory splitFactory,
+            boolean splittable)
+    {
+        List<Iterator<LocatedFileStatus>> fileStatusIterators = new ArrayList<>();
+        List<Path> paths = new ArrayList<>();
+
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(
+                hdfsEnvironment.doAs(hdfsContext.getIdentity().getUser(), () -> fileSystem.open(symlinkFile.getPath())), StandardCharsets.UTF_8))) {
+            CharStreams.readLines(reader).stream()
+                    .map(this::canonicalizeSymlinkPath)
+                    .forEach(paths::add);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_BAD_DATA, "Error parsing symlinks from: " + symlinkFile, e);
+        }
+
+        ImmutableSet<Path> parents = paths.stream()
```

Comment:
Could we be a bit more defensive with the optimization, e.g: not use it when `size(parents) ~= size(paths)`?

Reply:
Refactored to only use the optimization when `size(parents) = 1`